jwql/utils/utils.py

```diff
@@ -109,31 +94,108 @@ def filename_parser(filename):
     ValueError
         When the provided file does not follow naming conventions
     """
-    filename = os.path.basename(filename)
 
+    filename = os.path.basename(filename)
     file_root_name = (len(filename.split('.')) < 2)
 
-    regex_string_to_compile = r"[a-z]+" \
-                               "(?P<program_id>\d{5})"\
-                               "(?P<observation>\d{3})"\
-                               "(?P<visit>\d{3})"\
-                               "_(?P<visit_group>\d{2})"\
-                               "(?P<parallel_seq_id>\d{1})"\
-                               "(?P<activity>\w{2})"\
-                               "_(?P<exposure_id>\d+)"\
-                               "_(?P<detector>\w+)"
+    # Stage 1 and 2 filenames, e.g. "jw80500012009_01101_00012_nrcalong_uncal.fits"
+    stage_1_and_2 = \
+        r"jw" \
+        r"(?P<program_id>\d{5})"\
+        r"(?P<observation>\d{3})"\
+        r"(?P<visit>\d{3})"\
+        r"_(?P<visit_group>\d{2})"\
+        r"(?P<parallel_seq_id>\d{1})"\
+        r"(?P<activity>\w{2})"\
+        r"_(?P<exposure_id>\d+)"\
+        r"_(?P<detector>((?!_)[\w])+)"
+
+    # Stage 3 filenames with target ID, e.g. "jw80600-o009_t001_miri_f1130w_i2d.fits"
+    stage_3_target_id = \
+        r"jw" \
+        r"(?P<program_id>\d{5})"\
+        r"-(?P<ac_id>(o|c|a|r)\d{3})"\
```

Comment:
This also just made me realize that AC_IDs with `c`, `a`, or `r` can have 4 digits where the regex string is currently looking for 3. Also on lines 126, 135, and 145.

Reply:
@laurenmarietta 😱 Nice catch.  I'll try to get around this one too.  Perhaps there is a regex option where i can say "3 or 4".